File: OrangeJuice.Server.Core/Web/QueryStringNameValueCollection.cs
using System;
using System.Collections.Specialized;
using System.Linq;

namespace OrangeJuice.Server.Web
{
	public sealed class QueryStringNameValueCollection : NameValueCollection
	{
		#region Fields
		private readonly IUrlEncoder _urlEncoder;
		#endregion

		#region Ctor
		public QueryStringNameValueCollection(NameValueCollection coll, IUrlEncoder urlEncoder)
			: base(coll)
		{
			_urlEncoder = urlEncoder;
		}
		#endregion

		#region Methods
		public override string Get(string name)
		{
			return base.Get(name) ?? String.Empty;
		}

		public override string ToString()
		{
			var args = this.AllKeys
						   .Select(k => new { Key = k, Value = _urlEncoder.Encode(this[k]) })
						   .Select(p => String.Format("{0}={1}", p.Key, p.Value));
			return String.Join("&", args);
		}
		#endregion
	}
}

File: OrangeJuice.Server.Core.FSharp/Web/QueryStringNameValueCollection.fs
namespace OrangeJuice.Server.FSharp.Web

open System
open System.Collections.Generic
open System.Collections.Specialized
open System.Linq

open OrangeJuice.Server.Web
 
type QueryStringNameValueCollection(coll : NameValueCollection, urlEncoder : IUrlEncoder) =
    inherit NameValueCollection(coll)

    override this.Get(name : string) : string =
        let value = base.Get(name)
        if value <> null
            then value
            else String.Empty

    override this.ToString() : string =
        let args = this.AllKeys |>
                   Seq.map (fun k -> new KeyValuePair<string, string>(k, urlEncoder.Encode(this.Item(k)))) |>
                   Seq.map (fun p -> sprintf "%s=%s" p.Key p.Value)
        String.Join("&", args) 
